test/spmd/test_xla_distributed_checkpoint.py

```diff
@@ -0,0 +1,145 @@
+    return planner
+
+  def _same_shard_data(self, shards, others) -> bool:
+    for a, b in zip(shards, others):
+      if not torch.allclose(a.data, b.data):
+        return False
+    return True
+
+  def test_state_dict_separation(self):
+    model = self._get_sharded_model()
+    planner = self._get_load_planner(model)
+    if self.n_devices > 1:
+      # The state_dict should be flattened and separated
+      self.assertCountEqual(planner.sharded_state_dict, ['fc1.weight'])
+      # `fc2.weight` should be in the unsharded_state_dict despite having
+      # an explicit mark_sharding call because it is replicated.
+      self.assertCountEqual(planner.unsharded_state_dict,
+                            ['fc1.bias', 'fc2.weight', 'fc2.bias'])
+    else:
+      # With a single device, no tensors are sharded.
+      self.assertCountEqual(
+          planner.unsharded_state_dict,
+          ['fc1.weight', 'fc1.bias', 'fc2.weight', 'fc2.bias'])
+
+  def test_local_load_plan(self):
+    model = self._get_sharded_model()
+    planner = self._get_load_planner(model)
+    plan = planner.create_local_plan()
+    parameter_count = len(list(model.parameters()))
+    if self.n_devices > 1:
+      # When the model is sharded across devices, fc1.weight will result in
+      # self.n_devices ReadItems while all other tensors result in a single ReadItem.
+      sharded_read_items = [
+          ri for ri in plan.items if ri.dest_index.fqn == 'fc1.weight'
+      ]
+      self.assertEqual(self.n_devices, len(sharded_read_items))
+      # Every other parameter should have a single ReadItem
+      unsharded_read_items = set(plan.items) - set(sharded_read_items)
+      self.assertEqual(parameter_count - 1, len(unsharded_read_items))
+    else:
+      # If unsharded, there should be a single ReadItem per model parameter
+      self.assertEqual(parameter_count, len(plan.items))
+
+  @unittest.skipIf(
+      xr.global_device_count() == 1,
+      "Resolve and commit logic is only relevant for non-replicated tensors")
+  def test_resolve_and_commit_tensor(self):
+    model = self._get_sharded_model()
+    planner = self._get_load_planner(model)
+    plan = planner.create_local_plan()
+
+    xtensor = xs.wrap_if_sharded(model.fc1.weight)
+    old_shards = xtensor.local_shards
+    sharded_read_items = [
+        ri for ri in plan.items if ri.dest_index.fqn == 'fc1.weight'
+    ]
+    for read_item in sharded_read_items:
+      # Before all ReadItems have been processed, the tensor shards should not be updated
+      self.assertTrue(self._same_shard_data(xtensor.local_shards, old_shards))
+      tensor = planner.resolve_tensor(read_item)
+      planner.commit_tensor(read_item, -tensor)
+    # After all ReadItems are processed, the local_shards should reflect the new values
+    self.assertFalse(self._same_shard_data(xtensor.local_shards, old_shards))
+
+  def test_load_state_dict(self):
+    dist.init_process_group(
```

Comment:
Why do we need the process_group?

Reply:
`torch.distributed.checkpoint` requires a process group to coordinate merging the local plans into a global plan. Even though this is just a single instance, we still need to initialize the PG to use the `save_state_dict` and `load_state_dict` APIs.

Reply:
Hmmm, is this because we are saving the unsharded CPU tensor? If it's required all the time, I feel like it's not a good UX.

Reply:
A PG is always required, but I'll need to check if we can use the XLA backend - my first attempt to replace `gloo` with an `xla` group errored out when creating the global plan.

If we build a higher-level interface, this could be hidden this from the user.

Reply:
Most of our own distributed API does't require PG. So, maybe we should try keep the consistency. Do you know why they require PG?

Reply:
It's used to centrally coordinate the global plans, e.g. all worker plans are sent to the coordinator using collectives in the process group.

Actually, I see we have the option to run without a PG by setting `use_dist=False` in the `(save|load)_state_dict` functions, but we lose out on the global coordination. This shouldn't be an issue for the LoadPlanner, and I suspect we can make the SavePlanner work without global planning as well - we'll just need to track replica ranks and dedupe tensors locally.

Thanks for bringing this up Jiewen, I'll experiment with it and update the PR.

Reply:
Synced offline - when taking a checkpoint, we would lose the consistency provided by the process group (the coordinator waits until all workers are finished before writing metadata, which depends on a PG).

I'll use `no_dist=True` in the tests to simplify the test code, but we will still expect users to have a CPU process group when taking a checkpoint.